File: 022breadthFirstValues.js
const a = new Node('a');
const b = new Node('b');
const c = new Node('c');
const d = new Node('d');
const e = new Node('e');
const f = new Node('f');

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;

//      a
//    /   \
//   b     c
//  / \     \
// d   e     f

// no recursion for breadth first traversals
const breadthFirstValues = (root) => {
  // todo
  if(root === null) return [];

  const queue = [root];
  const result = [];

  while(queue.length > 0){
    const current = queue.shift();
    result.push(current.val);
    if(current.left) queue.push(current.left);
    if(current.right) queue.push(current.right);
  }
  return result;
};

breadthFirstValues(a);
//    -> ['a', 'b', 'c', 'd', 'e', 'f']

File: 022breadth_first_values.py
from collections import deque

 
class Node:
    def __init__(self, val):
      self.val = val
      self.left = None
      self.right = None




def breadth_first_values(root):
  if root is None:
    return []

  values = []
  queue = deque([root])

  while queue:
    current = queue.popleft() # using this will run in O(n) time or else it will run O(n^2)
    values.append(current.val)

    if current.left:
      queue.append(current.left)

    if current.right:
      queue.append(current.right)

  return values
